Read tailer offset before file stat when detecting file rotations

gh123man opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Some users have reported "phantom" rotations where the agent reports a file rotation even though the file did not rotate. This could lead to the agent re-ingesting the file incorrectly. 

The theory behind this fix relies on the following scenario to happen (with imaginary numbers for example): 

1. `DidRotate` gets called
2. `Stat` is called on the file and reports a size of `100` and assigned to `fileSize`.
3. Immediately following the `Stat` call `50` bytes are written to the file, and tailed by the agent.
4. `t.lastReadOffset.Load()` is called and returns `150` and is assigned to `lastReadOffset`
5. `fileSize < lastReadOffset` evaluates to `true` - but the file didn't actually rotate. 
6. The tailer is restarted from the beginning because the agent thinks there is a new file. 

With this fix, `lastReadOffset` is set before `Stat` is called. So if this same timing condition is hit such that byes are written after `lastReadOffset` is calculated, the result of `Stat` can only be larger than `lastReadOffset` unless the file was actually truncated. So `fileSize < lastReadOffset` will evaluate to `false` in these cases. 

We (believe to) have observed this very rarely under high load scenarios where a large volume of logs are being written to a file. 

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Bug fix
<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

The suspected bug is not easy to reproduce. However this change should have no functional impact under normal circumstances. So to QA this change test: 
- tail a file and write some logs
- rotate the file
- write logs to the new file

make sure rotation works as expected. 

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

hush-hush committed:
Starts splitting config helpers from the config package

As we continue to migrate the Agent code to components the many helpers
in pkg/config need to be moved somewhere else. The goal is to have a
simple API for the Config component while offering helpers on top of it.

The goal is to have helpers that can take either a "pkg/config:Config" or a
"comp/core/config:Component". This will facilitate the migration to
components.

For now the helpers are moved to pkg/config/utils. This is temporary as
we brainstorm the best place for them. Starting to split them from the
main config package is still useful.

hush-hush committed:
Update documentation wording for 'exclude_ec2_tags'

hush-hush committed:
Send 'agent_checks' payload 1 min after startup

The 'agent_checks' payload contains the status of every checks run by
the agent. It's sent every 10 min. We no longer wait 10 min for the first
payload to be sent. Instead we send the first payload after 1 min and
every 10 min after that.

hush-hush committed:
Adding timeout to the health command

hush-hush committed:
Remove duplicated Logger and Config name in Agent cmd

Similar to what the cluster agent does, we don't want to duplicate the
logger and config name in every command.

gh123man committed:
Read tailer offset before file stat.

gh123man committed:
reno

maycmlee reviewed (commented):
A couple of suggestions.

tbavelier committed:
Fix labels/annotations as tags feature on APIService KSM metrics (#16737)

sgnn7 committed:
[config] Update available sites in the config template

The current site list is out of date. This change ensures that we
have all the available ones identified.

ahmed-mez committed:
pkg/clusteragent/admission: support inject all libs via RC (#16653)

guyarb committed:
Update system_probe.py (#16713)

* Update system_probe.py

* Fixed lint

guyarb committed:
usm: remove gotos (#16625)

* usm: remove gotos

* Fixed cr comment

* Fixed cr comment

guyarb committed:
usm: protocol classification: Classify in the kernel only if the feature is supported (#16663)

* usm: protocol classification: Classify in the kernel only if the feature is supported

* Fixed CI

guyarb committed:
usm: performance: Massive improvement in encoder (#16497)

* usm: performance: Massive improvement in encoder

Improving 70% in allocations, 60% in total bytes, and 50% in runtime"

* Fixed cr comments

* Using sync.pool

safchain committed:
[CWS] fix profiles path (#16725)

paulcacheux committed:
[CWS] fix build tags generation in accessors/field handlers generation code (#16654)

* fix generation of build tags after copyright and code generated line

* cleanup copyrights ignores

guyarb committed:
usm: Improve connection creation to reduce allocations (#16621)

just-chillin committed:
[process-agent] Fix bug where disabled message is not shown (#16703)

* [process-agent] Fix bug where disabled message is not shown

mftoure committed:
[CWS][SEC-4301] Add security context to pods in workload metadata store (#16669)

* [WIP] Retrieve security context on pod's containers

* [WIP] Retrieve security context on pod's containers

* Retrieve security context on pod's containers

just-chillin committed:
[process-agent] Change sysprobe config missing error to info on Darwin (#16745)

Do not log error message due to missing system-probe.yaml file on dawin

safchain committed:
[CWS] add option to disable path resolution (#16564)

ganeshkumarsv committed:
Update CodeQL Bundle version (#16640)

* Update codeql-analysis.yml

* remove codeql config

* Delete codeql_config.yml

AliDatadog committed:
[Container Filter] Fix metrics exclusion for corechecks (#16697)

julien-lebot committed:
Fix APM telemetry casing & order (#16765)

Kitchen failures strictly unrelated to this PR

paulcacheux committed:
trivy: revert to require `v0.0.0-00010101000000-000000000000` for trivy (#16766)

paulcacheux committed:
bump `github.com/avast/retry-go/v4` and `github.com/cenkalti/backoff/v4` (#16768)

* bump `github.com/avast/retry-go/v4` to v4.3.4

* bump `github.com/cenkalti/backoff/v4` to v4.2.1

gbbr committed:
pkg/trace/api: fix flaky TestFlatten test (#16515)

hmahmood committed:
[system-probe] Reorg ebpf code for reuse with CO-RE network tracer (#16157)

hmahmood committed:
Add config for disabling/enabling netlink conntracker fallback (#16647)

safchain committed:
[CWS] remove runtime-security folder (#16763)

ahmed-mez committed:
pkg/trace/api: fix test lint (#16775)

lebauce committed:
Update fakeintake README.md (#16658)

gh123man committed:
Update releasenotes/notes/fix-file-rotation-edge-case-42cae3c391b5c85d.yaml

Co-authored-by: May Lee <mayl@alumni.cmu.edu>

safchain committed:
[CWS] fix macos build (#16777)

gh123man commented:
Oh my. I guess `git pull --rebase` was NOT the right thing to do.

Ill re-create this PR.

scottopell commented:
Just to add a reference here, this moved to https://github.com/DataDog/datadog-agent/pull/16780
